framework/sandstone.cpp

```diff
@@ -2611,70 +2610,93 @@ static int exec_mode_run(int argc, char **argv)
+            if (!(*eit).cores.empty()) {
+                if (first_it == topo.packages.end()) {
+                    first_it = eit;
+                }
+                last_it = eit;
+                run_cpus.add_package(*eit);
+            }
+        }
 
-        sApp->enabled_cpus = run_cpus;
-        do {
-            ret = run_tests_on_cpu_set(triage_tests, run_cpus);
-        } while (!ret && ++k < sApp->retest_count);
+        if (!run_cpus.empty()) {
+            sApp->enabled_cpus = run_cpus;
+            // should we run any tests if retest_count == 0??
+            // Keep the loop running only for SUCCESSes
+            int k = 0;
+            do {
+                ret = run_tests_on_cpu_set(triage_tests, run_cpus);
+            } while ((ret == EXIT_SUCCESS) && ++k < sApp->retest_count);
 
-        if (ret) ever_failed = true; /* we've seen a failure */
+            if (ret > EXIT_SUCCESS) ever_failed = true; /* we've seen a **FAILURE** */
 
-        if (!ret && ever_failed) {
-            // the last socket removed is the main suspect
-            result.push_back(disabled_sockets.at(disabled_sockets.size() - 1));
-            break;
+            if ((ret == EXIT_SUCCESS) && ever_failed) {
+                // the last socket removed is the main suspect
+                result.push_back((*removed_it).id);
+                break;
+            }
         }
-    }
 
-    if (ret) { // failed on the last socket as well, so it's the main suspect
-        // re-run on the first to make sure the last one is faulty
-        LogicalProcessorSet run_cpus = {};
-        int k = 0;
+        // find first non-empty package to "remove"
+        while (it != topo.packages.end()) {
+            if (!(*it).cores.empty()) {
+                removed_it = it;
+                it++;
+                break;
+            } else {
+                it++;
+            }
+        }
+    }
 
-        run_cpus.add_package(topo.packages.at(0));
-        sApp->enabled_cpus = run_cpus;
+    if (ret > EXIT_SUCCESS) {
+        if (first_it != last_it) {
+            // run on the first to make sure the last one is faulty
+            // if failed on the last socket, the first is the main suspect
+            LogicalProcessorSet run_cpus = {};
+            run_cpus.add_package(*first_it);
+            sApp->enabled_cpus = run_cpus;
 
-        do {
-            ret = run_tests_on_cpu_set(triage_tests, run_cpus);
-        } while (!ret && ++k < sApp->retest_count);
+            int k = 0;
+            do {
+                ret = run_tests_on_cpu_set(triage_tests, run_cpus);
+            } while ((ret == EXIT_SUCCESS) && (++k < sApp->retest_count));
 
-        if (!ret && ever_failed) {
-            result.push_back(disabled_sockets.at(disabled_sockets.size() - 1));
+            if (ret == EXIT_SUCCESS) {
+                result.push_back((*removed_it).id);
+            }
+        } else {
+            // only one package has been checked
```

Comment:
Is there any point in running triage for single-socket systems?

Reply:
The function returns immediately if there's just one socket.

Reply:
We can have n-sockets platform with only one socket being used. The condition from prologue were removed as are covered by this code. And, this is fair to add single package condition without test execution.